README.rst
Django Nginx Memcache
=====================
Provides a view decorator. Use it to cache content in Memcache for nginx to
retrieve.

Provides functions and signals to use to invalidate the cache, too.

The cache keys are hashed using an md5 of the the request path *without*
GET parameters,

In progress 
-----------

IMPORTANT: If this has anything listed here, the code must be treated as unstable, even if it's on a master branch

To do
-----

#. Add support for manually pushing a page to the cache via a function call - this will enable eager cacheing

Installation
------------

#. The usual pip or easy_install from `github <https://github.com/torchbox/django-nginx-memcache>`_::

    pip install -e git://github.com/torchbox/django-nginx-memcache#egg=django-nginx-memcache

#. Add ``nginx_memcache`` to your installed apps::

    INSTALLED_APPS = (
        # ...
        'nginx_memcache',
        # ...
    )

#. If you wish to use the DB-backed cached-page lookup (so that you know what pages may have been cached, and so may need invalidating), run `syncdb` to add the required table:

    ./manage.py syncdb nginx_memcache

#. Let your app/project know about the cache-invalidation signals by importing them in a module you *know* will always be loaded (eg the `__init__.py` for your core app)

    from nginx_memcache.signals import (
        handle_single_page_invalidation,
        handle_multiple_page_invalidation
    )

#. Set the default cache timeout::

    CACHE_NGINX = True
    CACHE_NGINX_TIME = 3600 * 24  # 1 day, in seconds
    
    # Default backend to use from settings.CACHES
    # May need to update the nginx conf if this is changed
    CACHE_NGINX_ALIAS = 'default'
    
    # Whether or not a DB-backed lookup table is useds 
    CACHE_NGINX_USE_LOOKUP_TABLE = True # default is True

#. Setup Memcached appropriately as described in `Django's cache framework docs <http://docs.djangoproject.com/en/dev/topics/cache/#memcached>`_.

#. Install Nginx with the `set_misc <https://github.com/agentzh/set-misc-nginx-module>`_ or `set_hash module <https://github.com/simpl/ngx_http_set_hash>`_. This is required to compute md5 cache keys from within Nginx. (See installing nginx below).

#. Configure Nginx for direct Memcached page retrieval, i.e::

    # Nginx host configuration for demosite. 
    #
    # Attempts to serve a page from memcache, falling
    # back to Django if it's not available 
                             
    upstream gunicorn_demosite {
        server 127.0.0.1:8003 fail_timeout=0;
    }

    server {
        listen 80 default_server;

        # Listen for all server names - lots of sites will be CNAMED
        # to this server, and we won't know what/which.

        server_name _;

        access_log /var/log/nginx/demosite.access.log;
        error_log /var/log/nginx/demosite.error.log;

        # Uncomment the following if you want to check your 
        # memcache keys during during development:
        # log_format hashedgeneratedkey $hash_key;
        # log_format realkey $memcached_key;
        # access_log  /var/log/nginx/keys.log  hashedgeneratedkey;
        # access_log  /var/log/nginx/keys.log  realkey;

        location /static/ {
            root /usr/local/django/demosite/;
        }

        location /media/ {
            root /usr/local/django/virtualenvs/demosite/lib/python2.7/site-packages/django/contrib/admin/;
        }

        location @gunicorn {
            # This is the standard config for serving Django via gunicorn                                                                                                                            
            root /usr/local/django/demosite/;

            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header Host $http_host;
            proxy_redirect off;

            if (!-f $request_filename) {
                proxy_pass http://gunicorn_demosite;
                break;
            }

            client_max_body_size 10m;
        }

        location @cache_miss {
            # Pass on the request to gunicorn, creating
            # a URI with the hostname as well as the path                                                                                                  
            # See the docs if $is_args$args is confusing

            set $caught_uri $http_host$uri$is_args$args;
            try_files $caught_uri @gunicorn;
        }

        location / {
            # By default, see if we can serve things from memcache.

            # Extract cache key args and cache key.                                                                                                                                                 
            if ($http_cookie ~* "pv=([^;]+)(?:;|$)") {
                set $page_version $1;
            }

            # If you are running multiple sites off the same server, 
            # the cache key to include the domain, too, which nginx
            # doesn't consider part of the $uri. (SJ: it ought to do, but doesn't)

            set_md5 $hash_key $http_host$uri&pv=$page_version;

            # make sure this matches the cache prefix and version config in the Django project settings
            set $django_cache_prefix ps;
            set $django_cache_version 1;

            set $memcached_key $django_cache_prefix:$django_cache_version:$hash_key;

            recursive_error_pages on;

            set $fallthrough_uri null;

            default_type       text/html;
            memcached_pass     127.0.0.1:11211;
            
            # We hand off all of these to @cache_miss and its descendent handlers.
            # The = means the handlers determine the error code, which is a Good Thing     

            error_page         401 = @cache_miss;
            error_page         403 = @cache_miss;
            error_page         404 = @cache_miss;
            error_page         405 = @cache_miss;
    
            # Note that it is not permitted to have a try_files in the same
            # location block as a memcache_pass
        }
    }

Installing Nginx
~~~~~~~~~~~~~~~~

These instructions apply for Ubuntu 11.04 and above::

    # install all dependencies
    sudo aptitude install libc6 libpcre3 libpcre3-dev libpcrecpp0 libssl0.9.8 libssl-dev zlib1g zlib1g-dev lsb-base

    # download nginx
    wget http://nginx.org/download/nginx-1.0.11.tar.gz
    tar -zxf nginx-1.0.11.tar.gz
    rm nginx-1.0.11.tar.gz
    cd nginx-1.0.11/

    # download modules
    wget https://github.com/simpl/ngx_devel_kit/zipball/v0.2.17 -O ngx_devel_kit.zip
    unzip ngx_devel_kit.zip
    wget https://github.com/agentzh/set-misc-nginx-module/zipball/v0.22rc4 -O set-misc-nginx-module.zip
    unzip set-misc-nginx-module.zip
    wget https://github.com/agentzh/echo-nginx-module/zipball/v0.37rc7 -O echo-nginx-module.zip
    unzip echo-nginx-module.zip

    # configure and install
    ./configure \
        --add-module=simpl-ngx_devel_kit-bc97eea \
        --add-module=agentzh-set-misc-nginx-module-290d6cb \
        --add-module=agentzh-echo-nginx-module-b7ea185 \
        --prefix=/usr \
        --pid-path=/var/run/nginx.pid \
        --lock-path=/var/lock/nginx.lock \
        --http-log-path=/var/log/nginx/access.log \
        --error-log-path=/var/log/nginx/error.log \
        --http-client-body-temp-path=/var/lib/nginx/body \
        --conf-path=/etc/nginx/nginx.conf \
        --with-http_flv_module \
        --with-http_ssl_module \
        --with-http_gzip_static_module \
        --http-proxy-temp-path=/var/lib/nginx/proxy \
        --with-http_stub_status_module \
        --http-fastcgi-temp-path=/var/lib/nginx/fastcgi \
        --http-uwsgi-temp-path=/var/lib/nginx/uwsgi \
        --http-scgi-temp-path=/var/lib/nginx/scgi
    make
    sudo make install

    # Done, now configure your nginx.


Usage
-----

nginx_memcache.decorators.cache_page_nginx
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The ``cache_page_nginx`` decorator caches the view's response content in Memcache. Any arguments are optional and outlined below.

Example::

    from nginx_memcache.decorators import cache_page_nginx

    @cache_page_nginx
    def my_view(request):
        ...

This will cache the view's response string in Memcache, and hereafter Nginx
will serve from Memcache directly, without hitting your Django server,
until the cache key expires.

Optional parameters
+++++++++++++++++++

``cache_timeout``
  Defaults to ``settings.CACHE_NGINX_TIME`` if not specified.

``page_version_fn``
  Use this to return a stringifiable version of the page, depending on the
  request. Example::

    def get_page_version(request):
        if request.user.is_authenticated():
            return 'authed'
        return 'anonymous'

``anonymous_only``
  Don't cache the page unless the user is anonymous, i.e. not authenticated.

Usage with forms and CSRF
~~~~~~~~~~~~~~~~~~~~~~~~~

If you want to embed forms on a cached page, you can leave out the context `{{ csrf() }}` or `{% csrf_token %}` and, instead, append it to all forms using JavaScript post page-load, or when a button is clicked.

Here's example JS and Django code for it::

    // JS code
    $.ajax({
        url: // your csrf url,
        type: 'GET',
        data: {type: 'login'},  // only if you need a session id for cookie login
        dataType: 'json',
        success: function(data) {
            $('form').each(function() {
                $(this).append(
                    '<input type=hidden name=csrfmiddlewaretoken ' +
                        ' value="' + data.token + '">');
            });
        }
    });

    // Django code
    # views.py, don't forget to add to urls.py
    def get_csrf(request):
        if request.GET.get('type') == 'login':
            request.session.set_test_cookie()
        return JSONResponse({
            'status': 1,
            'token': getattr(request, 'csrf_token', 'NOTPROVIDED')
        })


Full List of Settings
~~~~~~~~~~~~~~~~~~~~~

``CACHE_NGINX``
  Set this to False to disable any caching. E.g. for testing, staging...

``CACHE_NGINX_TIME``
  Default cache timeout.

``CACHE_NGINX_ALIAS``
  Which cache backend to use from `settings.CACHES <https://docs.djangoproject.com/en/dev/ref/settings/#std:setting-CACHES>`_

``CACHE_MINIFY_HTML``
  Will cache a HTML minified version of the response output. Default = False.

Contributing
============
If you'd like to fix a bug, add a feature, etc

#. Start by opening an issue.
    Be explicit so that project collaborators can understand and reproduce the
    issue, or decide whether the feature falls within the project's goals.
    Code examples can be useful, too.

#. File a pull request.
    You may write a prototype or suggested fix.

#. Check your code for errors, complaints.
    Use `check.py <https://github.com/jbalogh/check>`_

#. Write and run tests.
    Write your own test showing the issue has been resolved, or the feature
    works as intended.

Running Tests
=============
To run the tests::

    python manage.py test nginx_memcache


CHANGELOG.rst
Changelog
=========

0.1
-----
#. Initial release.


0.2-alpha
-----
#. Added support for generating cache keys based on full URIs, not just the path; this means the project can be used on shared servers, or servers serving multiple sites/subdomains which may have the same URI path

#. Added support for a database-backed lookup table of cached pages, to make mass-page invalidation easier

#. Added support for bulk invalidation of pages associated with a particular hostname (and/or subset of those pages. eg: if Cached Page A is hyperlinked from a menu that features on all pages in the site and its title changes, you need to invalidate more than just A, but - particularly if you're serving/cacheing multiple sites - you don't want to blat the entire nginx cache.)

#. Added support for invalidation of cached pages via a signal. Applies to both single pages and bunches of pages with the same `lookup_identifier` or `supplementary_identifier` 

nginx_memcache/tests/signals.py
"""Feature-level tests to confirm that
the signal handlers are working as expected
"""

from django.http import HttpResponse
from django.test import TestCase
from django.test.client import RequestFactory
from django.conf import settings

from nginx_memcache.decorators import cache_page_nginx
from nginx_memcache.models import CachedPageRecord
from nginx_memcache.cache import (
    nginx_cache as cache,
    get_cache_key,
    CACHE_NGINX_DEFAULT_COOKIE
)
from nginx_memcache.signals import (
    invalidate_single_page,
    invalidate_many_pages
)

class CacheSignalTests(TestCase):

    def setUp(self):
        #monkey-patch settings for test purposes
        setattr(settings, 'CACHE_NGINX_USE_LOOKUP_TABLE', True)

        self.factory = RequestFactory()
        # Clear the cache before we do anything.
        self.request = self.factory.get('/', SERVER_NAME="example1.com")
        cache.clear()
        self.cache_key = get_cache_key(
            self.request.get_host(),
            self.request.get_full_path()
        )
        assert not cache.get(self.cache_key)
        self.assertEqual(CachedPageRecord.objects.count(), 0)

    def my_view(self, request):
        """Class-wide test view"""
        return HttpResponse('content')

    def test_invalidate_single_page(self):

        # Cache a page
        my_view_cached = cache_page_nginx(self.my_view)
        self.assertEqual(my_view_cached(self.request).content, 'content')

        # Confirm in cache
        assert cache.get(self.cache_key)

        # invalidate the cache via the signal
        invalidate_single_page.send_robust(
            sender=None,  # Not relevant in our handler code
            request_host=u"example1.com",
            request_path=u"/",
            page_version='',
            cookie_name=CACHE_NGINX_DEFAULT_COOKIE
        )
        # confirm no longer in cache
        assert not cache.get(self.cache_key)

    def test_invalidate_single_page_using_minimal_args(self):
        # Also confirm that invalidation doesn't need
        # page_version or cookie_name kwargs

        # Cache a page
        my_view_cached = cache_page_nginx(self.my_view)
        self.assertEqual(my_view_cached(self.request).content, 'content')

        # Confirm in cache
        assert cache.get(self.cache_key)

        # invalidate the cache via the signal
        invalidate_single_page.send_robust(
            sender=None,  # Not relevant in our handler code
            request_host=u"example1.com",
            request_path=u"/",
        )
        # confirm no longer in cache
        assert not cache.get(self.cache_key)

    def test_invalidate_many_pages(self):

        # 1. Cache two pages with the same lookup_identifier

        # Put two pages into the cache, using the same view, but two separate
        # paths, so they count as separate pages

        request_1 = self.factory.get('/foo/', SERVER_NAME="example1.com")
        request_1_cache_key = get_cache_key(
            request_1.get_host(),
            request_1.get_full_path()
        )
        my_view_cached = cache_page_nginx(
            self.my_view,
            lookup_identifier="test1"
        )
        self.assertEqual(my_view_cached(request_1).content, 'content')

        request_2 = self.factory.get('/bar/', SERVER_NAME="example1.com")
        request_2_cache_key = get_cache_key(
            request_2.get_host(),
            request_2.get_full_path()
        )
        my_view_cached = cache_page_nginx(
            self.my_view,
            lookup_identifier="test1"
        )
        self.assertEqual(my_view_cached(request_2).content, 'content')

        # 2. Cache two pages with a different lookup_identifier
        #    from above, but unique supplementary_identifiers

        request_3 = self.factory.get('/moo/', SERVER_NAME="example1.com")
        request_3_cache_key = get_cache_key(
            request_3.get_host(),
            request_3.get_full_path()
        )
        my_view_cached = cache_page_nginx(
            self.my_view,
            lookup_identifier="test2",
            supplementary_identifier="AAA"
        )
        self.assertEqual(my_view_cached(request_3).content, 'content')

        request_4 = self.factory.get('/boo/', SERVER_NAME="example1.com")
        request_4_cache_key = get_cache_key(
            request_4.get_host(),
            request_4.get_full_path()
        )
        my_view_cached = cache_page_nginx(
            self.my_view,
            lookup_identifier="test2",
            supplementary_identifier="BBB"
        )
        self.assertEqual(my_view_cached(request_4).content, 'content')

        # 3. Confirm all in cache and that keys are unique

        assert cache.get(request_1_cache_key)
        assert cache.get(request_2_cache_key)
        assert cache.get(request_3_cache_key)
        assert cache.get(request_4_cache_key)

        self.assert_(
            request_1_cache_key != request_2_cache_key != \
            request_3_cache_key != request_4_cache_key
        )

        # 4. Invalidate first lookup_identifier via the signal
        invalidate_many_pages.send_robust(
            sender=None,  # Not needed
            lookup_identifier="test1"
        )

        # 5. Confirm those two pages no longer in cache, but
        #    the others are
        assert not cache.get(request_1_cache_key)
        assert not cache.get(request_2_cache_key)
        assert cache.get(request_3_cache_key)
        assert cache.get(request_4_cache_key)

        # 6. Invalidate using the second
        #    lookup_identifier + first supplementary_identifier
        #    via the signal
        invalidate_many_pages.send_robust(
            sender=None,  # Not needed
            lookup_identifier="test2",
            supplementary_identifier="AAA"
        )

        # 7. Confirm that single page no longer in cache
        assert not cache.get(request_1_cache_key)
        assert not cache.get(request_2_cache_key)
        assert not cache.get(request_3_cache_key)
        assert cache.get(request_4_cache_key)

        # 8. invalidate using the second
        #    lookup_identifier as a blanket call

        invalidate_many_pages.send_robust(
            sender=None,  # Not needed
            lookup_identifier="test2"
        )

        # 9. confirm that the final page no longer in cache
        assert not cache.get(request_1_cache_key)
        assert not cache.get(request_2_cache_key)
        assert not cache.get(request_3_cache_key)
        assert not cache.get(request_4_cache_key)


nginx_memcache/tests/__init__.py
from .decorators import CachePageDecoratorTests
from .cache import CachedPageRecordTests
from .signals import CacheSignalTests

nginx_memcache/models.py
from django.conf import settings
from django.db import models


class CachedPageRecord(models.Model):
    """When a page is cached, the user has the option (via the decorator) to
    note this in a DB-backed lookup table. This data is stored in this model.

    Note that you'll ideally be using a database implementation
    that can enforce UNIQUE constraints.

    The most interesting use of this model is cache.bulk_invalidate() function.
    Read its docstring for more info.

    """

    # NB: because the base_cache_key is the primary key,
    # there is no 'id' field on this model
    base_cache_key = models.CharField(
        max_length=32,  # standard output of hexdigest() on an md5
        primary_key=True,  # which implies unique=True
        help_text=(
            "The md5 based on the host and path. This will be " +
            "combined with any prefix and the version to get the " +
            "real cache key stored in memcache"
        )
    )

    parent_identifier = models.CharField(
        blank=False,
        null=False,
        max_length=255,
        db_index=True,
        help_text=(
            "A 255-char identifier for whatever your parent object is." +
            "Could be a slug, a subdomain, or a Site ID [as a string]. " +
            "If no lookup_identifier is passed to the cache decorator, " +
            "this value will be the hostname of the server (which is fine)."
        )
    )

    supplementary_identifier = models.CharField(
        blank=True,
        null=True,
        max_length=45,  # kept short to avoid breaking MySQL's 1000-byte index limit
        help_text=(
            "Additional meta string you can use to scope your invalidation " +
            "calls to a subset of keys - eg 'news' or 'category_3'. 45 chars max."
        )
    )

    class Meta:
        unique_together = (
            (
                'parent_identifier',
                'base_cache_key',
                'supplementary_identifier'
            ),
        )

    def __unicode__(self):
        return "%s/%s/%s" % (
            self.base_cache_key,
            self.parent_identifier,
            self.supplementary_identifier
        )

    @property
    def memcached_key(self):
        """
        Complete representation of the cache key used by memcache.
        Includes the appropriate Django prefix and version.

        """

        CACHE_ALIAS = getattr(settings, 'CACHE_NGINX_ALIAS', 'default')
        return "%s:%s:%s" % (
            settings.CACHES[CACHE_ALIAS].get('KEY_PREFIX'),
            1,  # CACHE_VERSION defaults to 1 and nginx only ever seeks that
            self.base_cache_key
        )


nginx_memcache/signals.py
"""Custom signals and their handlers,
to make it easier to call the
invalidation functions
"""

from django.dispatch import Signal, receiver

from .cache import invalidate, bulk_invalidate

# Signals
invalidate_single_page = Signal(
    providing_args=[
        "request_host",
        "request_path",
        "page_version",
        "cookie_name"
    ]
)

invalidate_many_pages = Signal(
    providing_args=[
        "lookup_identifier",
        "supplementary_identifier",
    ]
)

# Handlers, connected to those signals 

@receiver(invalidate_single_page)
def handle_single_page_invalidation(sender, signal, **provided_args):
    invalidate(**provided_args)  # Hand it on with just the core things in there

@receiver(invalidate_many_pages)
def handle_multiple_page_invalidation(sender, signal, **provided_args):
    bulk_invalidate(**provided_args)  


nginx_memcache/decorators.py
from django.conf import settings
from django.utils.decorators import decorator_from_middleware_with_args

from .middleware import UpdateCacheMiddleware

# Create decorator.
# This is a similar approach to https://github.com/shaunsephton/djanginxed
# except using middleware instead of writing decorator directly.
def cache_page_nginx(
        view_fn=None,
        cache_timeout=settings.CACHE_NGINX_TIME,
        page_version_fn=None,
        anonymous_only=False,
        lookup_identifier=None,
        supplementary_identifier=None
    ):
    decorator = decorator_from_middleware_with_args(UpdateCacheMiddleware)(
        cache_timeout=cache_timeout,
        page_version_fn=page_version_fn,
        anonymous_only=anonymous_only,
        lookup_identifier=lookup_identifier,
        supplementary_identifier=supplementary_identifier
    )
    if callable(view_fn):
        return decorator(view_fn)
    return decorator


nginx_memcache/middleware.py
from django.conf import settings

from .cache import cache_response


is_enabled = getattr(settings, 'CACHE_NGINX', True)


class UpdateCacheMiddleware(object):
    """Updates the cache cache with the response of the request.

    It is of _paramount_ importance that the generated cache_key matches
    exactly the key generated by your web server (nginx) to lookup the page
    from the cache.

    Remember to set CACHE_NGINX_ALIAS to a cache backend that uses memcache.

    The middleware must be at the top of settings.MIDDLEWARE_CLASSES
    to be called last during the response phase.

    """

    def __init__(
            self,
            cache_timeout,
            page_version_fn,
            anonymous_only,
            lookup_identifier=None,
            supplementary_identifier=None
        ):
        """Initialize middleware. Args:
            * cache_timeout - seconds after which the cached response expires
            * page_version_fn - return a value to version the view based on
                    the request.
            * anonymous_only - only cache if the user is anonymous
            * lookup_identifer - for populating the lookup table;
                see models.CachedPageRecord. If not specified in the decorator,
                the value of request.get_host() will be used, which is a good
                thing to use anyway.
            * supplementary_identifier - entirely optional scoping variable.
                For populating the lookup table; see models.CachedPageRecord

        """

        self.cache_timeout = cache_timeout
        self.page_version_fn = page_version_fn
        self.anonymous_only = anonymous_only
        self.lookup_identifier = lookup_identifier
        self.supplementary_identifier = supplementary_identifier

    def process_response(self, request, response):
        """Sets the cache, if needed."""
        if not is_enabled or request.method != 'GET' or (
            response.status_code != 200):
            # HTTPMiddleware, throws the body of a HEAD-request away before
            # this middleware gets a chance to cache it.
            return response

        # Logged in users don't cause caching if anonymous_only is set.
        if self.anonymous_only and request.user.is_authenticated():
            return response

        # Otherwise, we do want to cache the response.
        cache_response(
            request,
            response,
            cache_timeout=self.cache_timeout,
            page_version_fn=self.page_version_fn,
            lookup_identifier=self.lookup_identifier,
            supplementary_identifier=self.supplementary_identifier
        )
        return response


nginx_memcache/cache.py
import hashlib

from django.conf import settings
from django.core.cache import get_cache
from django.db import IntegrityError
from django.template.response import TemplateResponse
from django.utils.encoding import DjangoUnicodeDecodeError
from django.utils.html import strip_spaces_between_tags as minify_html

from .models import CachedPageRecord

CACHE_NGINX_DEFAULT_COOKIE = getattr(settings, 'CACHE_NGINX_COOKIE', 'pv')
CACHE_TIME = getattr(settings, 'CACHE_NGINX_TIME', 3600 * 24)
CACHE_ALIAS = getattr(settings, 'CACHE_NGINX_ALIAS', 'default')
CACHE_MINIFY_HTML = getattr(settings, 'CACHE_MINIFY_HTML', False)
nginx_cache = get_cache(CACHE_ALIAS)

def cache_response(
        request,
        response,
        cache_timeout=CACHE_TIME,
        cookie_name=CACHE_NGINX_DEFAULT_COOKIE,
        page_version_fn=None,
        lookup_identifier=None,
        supplementary_identifier=None
    ):

    """Class based view responses TemplateResponse objects and do not call
    render automatically, you we must trigger this."""
    if type(response) is TemplateResponse and not response.is_rendered:
        response.render()

    """ Minify the HTML outout if set in settings. """
    if CACHE_MINIFY_HTML:
        if 'text/html' in response['Content-Type']:
            try:
                response.content = minify_html(response.content.strip())
            except DjangoUnicodeDecodeError:
                pass

    """Cache this response for the web server to grab next time."""
    # get page version
    if page_version_fn:
        pv = page_version_fn(request)
    else:
        pv = ''
    cache_key = get_cache_key(
        request_host=request.get_host(),
        request_path=request.get_full_path(),
        page_version=pv,
        cookie_name=cookie_name
    )
    nginx_cache.set(cache_key, response._get_content(), cache_timeout)

    # Store the version, if any specified.
    if pv:
        response.set_cookie(cookie_name, pv)

    # Add record of cacheing taking place to
    # invalidation lookup table, if appropriate
    if getattr(settings, 'CACHE_NGINX_USE_LOOKUP_TABLE', False):
        if not lookup_identifier:
            # If no identifier specified, use the hostname.
            # If you prefer, you could pass in a Site.id, etc
            lookup_identifier = request.get_host()
        add_key_to_lookup(
            cache_key,
            lookup_identifier,
            supplementary_identifier
        )


def get_cache_key(
        request_host,
        request_path,
        page_version='',
        cookie_name=CACHE_NGINX_DEFAULT_COOKIE
    ):
    """ Use the request host, request path and
        optional page version to get cache key."""
    raw_key = u'%s%s&%s=%s' % (
        request_host,
        request_path,
        cookie_name,
        page_version
    )
    return hashlib.md5(raw_key).hexdigest()


def invalidate_from_request(
        request,
        page_version='',
        cookie_name=CACHE_NGINX_DEFAULT_COOKIE
    ):
    """Delete cache key for this request and page version."""
    invalidate(
        request_host=request.get_host(),
        request_path=request.get_full_path(),
        page_version=page_version,
        cookie_name=cookie_name
    )


def invalidate(
        request_host,
        request_path,
        page_version='',
        cookie_name=CACHE_NGINX_DEFAULT_COOKIE
    ):
    """Delete cache key for this request path and page version."""
    cache_key = get_cache_key(
        request_host=request_host,
        request_path=request_path,
        page_version=page_version,
        cookie_name=cookie_name
    )
    nginx_cache.delete(cache_key)


def bulk_invalidate(
        lookup_identifier,
        supplementary_identifier=None
    ):
    """Find all the pages in the lookup table that are identifed by the args
    and invalidate the/any cache for them.

    Note that this does not invalidate by URI-derived key; it invalidates
    *all* URI-derived keys via the particular lookup_identifier (and an
    optional supplementary_identifier) they have been associated with.

    Unless explicitly set in the nginx_cache_page decorator, the
    lookup_identifier is the value of request.get_host()

    Examples: a page with Django-generated cache key of
    'aaabbaaabbbaabbababab2424242412' may be within a site foo.com.

    So, calling bulk_invalidate with the output of request.get_host() as
    the layout_identifier argument will cause any cached paged for that host
    to be invalidated.

    And if the page with a key of 'aaabbaaabbbaabbababab2424242412' was
    within a 'news' section of site, and links to it may appear on other
    'news' pages, but nowhere else in the site, you may have a use case where
    only 'news' pages really needed to be invalidated.

    So, provided that the supplementary_identifier arg 'news' was passed to the
    nginx_cache_page decorator in the first place, you can invalidate only the
    pages in that 'news' subset by passing the 'news' as the
    supplementary_identifier here.

    """

    relevant_records = CachedPageRecord.objects.filter(
        parent_identifier=lookup_identifier,
    )

    if supplementary_identifier:
        relevant_records = relevant_records.filter(
            supplementary_identifier=supplementary_identifier
        )

    nginx_cache.delete_many(
        [record.base_cache_key for record in relevant_records]
    )
    # NB: we _don't_ delete the objects for the keys we've just invalidated -
    # there's little overhead in trying to invalidate an already-invalid key
    # in memcache, whereas dropping rows from the DB to replace them
    # potentially milliseconds later is a more significant hit.
    #
    # Yes, in the future, there will be some kind of cleanup, probably
    # introducing an expiry_datetime field on the model


def add_key_to_lookup(
        cache_key,
        lookup_identifier,
        supplementary_identifier
    ):
    """Adds a CachedPageRecord to the lookup table, ensuring no duplicates of
       this data are also stored.
    """

    cpr = CachedPageRecord(
        base_cache_key=cache_key,
        parent_identifier=lookup_identifier,
        supplementary_identifier=supplementary_identifier
    )
    try:
        cpr.save()
    except IntegrityError:
        # Because it already exists, and we only
        # ever need one entry per cached page
        pass


def remove_key_from_lookup(
        cache_key,
        lookup_identifier,
        supplementary_identifier
    ):
    """Not currently tested, but will cleanly remove a CachedPageRecord"""

    try:
        cpr = CachedPageRecord.objects.get(
            base_cache_key=cache_key,
            parent_identifier=lookup_identifier,
            supplementary_identifier=supplementary_identifier
        )
        cpr.delete()
    except CachedPageRecord.DoesNotExist:
        pass
